### PR TITLE
Update metric changes

### PR DESCRIPTION
**Pull request by yiheng-wang-nv**
Some of the tutorials should be updated due to the changes for metric calculation. I have modified all related notebooks and py files, and correct a small typo in `brats_training_ddp.py`.

### CODE REVIEW

File: federated_learning/substra/assets/objective/metrics.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 
 class MonaiMetrics(tools.Metrics):
-    dice_metric = DiceMetric(include_background=True, to_onehot_y=False, sigmoid=True, reduction="mean")
+    dice_metric = DiceMetric(include_background=True, reduction="mean")
```
Comment (Nic-Ma) [inline federated_learning/substra/assets/objective/metrics.py:8]: Maybe you should also add `post_trans = Compose([Activations(sigmoid=True), AsDiscrete(threshold_values=True)])
    # try to use all the available GPUs` in this FL example?

File: 3d_segmentation/brats_segmentation_3d.ipynb [hunk 2]
```diff
@@ -478,24 +481,25 @@
     "                    val_data[\"label\"].to(device),\n",
     "                )\n",
     "                val_outputs = model(val_inputs)\n",
+    "                val_outputs = post_trans(val_outputs)\n",
     "                # compute overall mean dice\n",
-    "                value = dice_metric(y_pred=val_outputs, y=val_labels)\n",
-    "                not_nans = dice_metric.not_nans.item()\n",
+    "                value, not_nans = dice_metric(y_pred=val_outputs, y=val_labels)\n",
+    "                not_nans= not_nans.item()\n",
```
Comment (Nic-Ma) [inline 3d_segmentation/brats_segmentation_3d.ipynb:487]: Missing a blank before "=".

File: 3d_segmentation/ignite/unet_evaluation_array.py [hunk 3]
```diff
@@ -66,17 +66,20 @@ def main(tempdir):
...
     sw_batch_size = 4
 
+    post_trans = Compose([Activations(sigmoid=True), AsDiscrete(threshold_values=True)])
+
     def _sliding_window_processor(engine, batch):
         net.eval()
         with torch.no_grad():
             val_images, val_labels = batch[0].to(device), batch[1].to(device)
             seg_probs = sliding_window_inference(val_images, roi_size, sw_batch_size, net)
+            seg_probs = post_trans(seg_probs)
```
Comment (Nic-Ma) [inline 3d_segmentation/ignite/unet_evaluation_array.py:76]: As the output is already through `post_transform`, no need to call `predict_segmentation()` in SegmentationSaver anymore, please update all the evaluation examples if didn't delete it.
Thanks.

File: 3d_segmentation/ignite/unet_training_array.py [hunk 4]
```diff
@@ -112,11 +112,14 @@ def main(tempdir):
...
-    val_metrics = {metric_name: MeanDice(sigmoid=True, to_onehot_y=False)}
+    val_metrics = {metric_name: MeanDice()}
+
+    post_pred = Compose([Activations(sigmoid=True), AsDiscrete(threshold_values=True)])
+    post_label = AsDiscrete(threshold_values=True)
 
     # Ignite evaluator expects batch=(img, seg) and returns output=(y_pred, y) at every iteration,
     # user can add output_transform to return other values
-    evaluator = create_supervised_evaluator(net, val_metrics, device, True)
+    evaluator = create_supervised_evaluator(net, val_metrics, device, True, output_transform=lambda x, y, y_pred: (post_pred(y_pred), post_label(y)))
```
Comment (Nic-Ma) [inline 3d_segmentation/ignite/unet_training_array.py:122]: This line exceeds 120 chars.

File: 3d_segmentation/ignite/unet_training_dict.py [hunk 5]
```diff
@@ -152,11 +154,14 @@ def prepare_batch(batch, device=None, non_blocking=False):
...
-    val_metrics = {metric_name: MeanDice(sigmoid=True, to_onehot_y=False)}
+    val_metrics = {metric_name: MeanDice()}
+
+    post_pred = Compose([Activations(sigmoid=True), AsDiscrete(threshold_values=True)])
+    post_label = AsDiscrete(threshold_values=True)
 
     # Ignite evaluator expects batch=(img, seg) and returns output=(y_pred, y) at every iteration,
     # user can add output_transform to return other values
-    evaluator = create_supervised_evaluator(net, val_metrics, device, True, prepare_batch=prepare_batch)
+    evaluator = create_supervised_evaluator(net, val_metrics, device, True, output_transform=lambda x, y, y_pred: (post_pred(y_pred), post_label(y)), prepare_batch=prepare_batch)
```
Comment (Nic-Ma) [inline 3d_segmentation/ignite/unet_training_dict.py:164]: This line exceeds 120 chars.

File: 3d_segmentation/torch/unet_evaluation_array.py [hunk 6]
```diff
@@ -75,7 +75,8 @@ def main(tempdir):
             roi_size = (96, 96, 96)
             sw_batch_size = 4
             val_outputs = sliding_window_inference(val_images, roi_size, sw_batch_size, model)
-            value = dice_metric(y_pred=val_outputs, y=val_labels)
+            val_outputs = post_trans(val_outputs)
+            value, _ = dice_metric(y_pred=val_outputs, y=val_labels)
             metric_count += len(value)
             metric_sum += value.item() * len(value)
             val_outputs = (val_outputs.sigmoid() >= 0.5).float()
```
Comment (Nic-Ma) [inline 3d_segmentation/torch/unet_evaluation_array.py:82]: Please delete this line, it's not necessary anymore.
Thanks.

File: 3d_segmentation/torch/unet_evaluation_dict.py [hunk 7]
```diff
@@ -89,7 +89,8 @@ def main(tempdir):
             roi_size = (96, 96, 96)
             sw_batch_size = 4
             val_outputs = sliding_window_inference(val_images, roi_size, sw_batch_size, model)
-            value = dice_metric(y_pred=val_outputs, y=val_labels)
+            val_outputs = post_trans(val_outputs)
+            value, _ = dice_metric(y_pred=val_outputs, y=val_labels)
             metric_count += len(value)
             metric_sum += value.item() * len(value)
             val_outputs = (val_outputs.sigmoid() >= 0.5).float()
```
Comment (Nic-Ma) [inline 3d_segmentation/torch/unet_evaluation_dict.py:96]: Please delete this line, it's not necessary anymore.
Thanks.

File: 3d_segmentation/unet_segmentation_3d_ignite.ipynb [hunk 8]
```diff
@@ -334,11 +336,13 @@
...
-    "val_metrics = {metric_name: MeanDice(sigmoid=True, to_onehot_y=False)}\n",
-    "\n",
+    "val_metrics = {metric_name: MeanDice()}\n",
+    "post_pred = Compose([Activations(sigmoid=True), AsDiscrete(threshold_values=True)])\n",
+    "post_label = AsDiscrete(threshold_values=True)\n",
     "# Ignite evaluator expects batch=(img, seg) and returns output=(y_pred, y) at every iteration,\n",
     "# user can add output_transform to return other values\n",
-    "evaluator = ignite.engine.create_supervised_evaluator(net, val_metrics, device, True)\n",
+    "evaluator = ignite.engine.create_supervised_evaluator(net, val_metrics, device, True, \\\n",
+    "                                                     output_transform=lambda x, y, y_pred: (post_pred(y_pred), post_label(y)))\n",
```
Comment (Nic-Ma) [inline 3d_segmentation/unet_segmentation_3d_ignite.ipynb:345]: This line seems too long.

File: 3d_segmentation/unet_segmentation_3d_ignite.ipynb [hunk 9]
```diff
@@ -94,7 +94,9 @@
     "from monai.networks import predict_segmentation\n",
```
Comment (Nic-Ma) [inline 3d_segmentation/unet_segmentation_3d_ignite.ipynb:94]: Please remove the usage of `predict_segmentation` in this file.
Thanks.

File: acceleration/distributed_training/brats_training_ddp.py [hunk 10]
```diff
@@ -141,7 +143,7 @@ def _generate_data_list(self, dataset_dir):
         data = super()._generate_data_list(dataset_dir)
         return partition_dataset(
             data=data,
-            num_partitons=dist.get_world_size(),
+            num_partitions=dist.get_world_size(),
```
Comment (Nic-Ma) [inline acceleration/distributed_training/brats_training_ddp.py:146]: Good finding! Thanks.

**Review by Nic-Ma — COMMENTED** [review]
Hi @yiheng-wang-nv ,

Thanks for your overall fix for the examples.
Please update the minor issues I put inline.

Thanks.

**Review by Nic-Ma — APPROVED** [review]
Looks good to me now.
Thanks.
